agent/config/builder.go

```diff
@@ -1113,12 +1132,50 @@ func (b *Builder) Validate(rt RuntimeConfig) error {
 		return fmt.Errorf("data_dir cannot be empty")
 	}
 
-	if !validContentPath.MatchString(rt.UIContentPath) {
-		return fmt.Errorf("ui-content-path can only contain alphanumeric, -, _, or /. received: %s", rt.UIContentPath)
+	if !validContentPath.MatchString(rt.UIConfig.ContentPath) {
+		return fmt.Errorf("ui-content-path can only contain alphanumeric, -, _, or /. received: %q", rt.UIConfig.ContentPath)
+	}
+
+	if hasVersion.MatchString(rt.UIConfig.ContentPath) {
+		return fmt.Errorf("ui-content-path cannot have 'v[0-9]'. received: %q", rt.UIConfig.ContentPath)
 	}
 
-	if hasVersion.MatchString(rt.UIContentPath) {
-		return fmt.Errorf("ui-content-path cannot have 'v[0-9]'. received: %s", rt.UIContentPath)
+	if rt.UIConfig.MetricsProvider != "" &&
+		!reBasicName.MatchString(rt.UIConfig.MetricsProvider) {
+		return fmt.Errorf("ui_config.metrics_provider can only contain lowercase "+
+			"alphanumeric or _ characters. received: %q", rt.UIConfig.MetricsProvider)
+	}
+	if rt.UIConfig.MetricsProviderOptionsJSON != "" {
+		// Attempt to parse the JSON to ensure it's valid, parsing into a map
+		// ensures we get an object.
+		var dummyMap map[string]interface{}
+		err := json.Unmarshal([]byte(rt.UIConfig.MetricsProviderOptionsJSON),
```

Comment:
We weren't happy with inlined JSON for connect escape hatches. Is this a similar idea? should we make this opaque config instead and serialize it to the UI?

Reply:
I considered this during RFC and I'm OK with this as it is for now - it's a highly experimental interface. Somewhat unlike the escape hatches where the JSON is protobuf we decode and merge, this is literally passed through to the front end which passes it directly back to the custom code the operator is using in JavaScript as a parameter. The decode here is just to try to ensure valid formatting so it doesn't just silently break the whole UI if there's a typo.

The UX of writing explicit JSON for a value that is passed to your javascript code doesn't seem as bad to me as writing raw JSON encoded proto3 Envoy config. The complexities of conversions between HCL and JSON seem to potentially make the UX more confusing?

That said, I'm open to iterating on this if it proves its value but I'm not sure the extra complexity of opaque config makes sense for now either from UX or from a "getting an MVP out and seeing if it's useful" perspective?